netty/src/main/java/io/grpc/netty/GrpcHttp2Headers.java

```diff
@@ -0,0 +1,154 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.netty;
+
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.util.AsciiString;
+
+import java.util.Iterator;
+import java.util.Map.Entry;
+
+import javax.annotation.Nullable;
+
+/**
+ * A custom implementation of Http2Headers that only includes methods used by gRPC.
+ */
+final class GrpcHttp2Headers extends AbstractHttp2Headers {
+
+  private final AsciiString[] normalHeaders;
+  private final AsciiString[] psuedoHeaders;
+  private final int psuedoHeadersCount;
+
+  /**
+   * Client constructor.
+   */
+  GrpcHttp2Headers(byte[][] serializedMetadata, @Nullable AsciiString authority,
+      @Nullable AsciiString path, @Nullable AsciiString method, @Nullable AsciiString scheme,
+      @Nullable AsciiString key1, @Nullable AsciiString value1, @Nullable AsciiString key2,
+      @Nullable AsciiString value2, @Nullable AsciiString key3, @Nullable AsciiString value3) {
+    normalHeaders = new AsciiString[serializedMetadata.length];
+    for (int i = 0; i < normalHeaders.length; i++) {
+      normalHeaders[i] = new AsciiString(serializedMetadata[i], false);
+    }
+    psuedoHeaders = new AsciiString[14]; // Number of extra headers * 2
+    int count = 0;
+    count = addPseudoHeader(Http2Headers.PseudoHeaderName.AUTHORITY.value(), authority, count);
+    count = addPseudoHeader(Http2Headers.PseudoHeaderName.PATH.value(), path, count);
+    count = addPseudoHeader(Http2Headers.PseudoHeaderName.METHOD.value(), method, count);
+    count = addPseudoHeader(Http2Headers.PseudoHeaderName.SCHEME.value(), scheme, count);
+    count = addPseudoHeader(key1, value1, count);
+    count = addPseudoHeader(key2, value2, count);
+    count = addPseudoHeader(key3, value3, count);
+    psuedoHeadersCount = count;
+  }
+
+  /**
+   * Server constructor.
+   */
+  GrpcHttp2Headers(byte[][] serializedMetadata, @Nullable AsciiString status,
+      @Nullable AsciiString key1, @Nullable AsciiString value1) {
+    normalHeaders = new AsciiString[serializedMetadata.length];
+    for (int i = 0; i < normalHeaders.length; i++) {
+      normalHeaders[i] = new AsciiString(serializedMetadata[i], false);
+    }
+    psuedoHeaders = new AsciiString[4]; // Number of extra headers * 2
+    int count = 0;
+    count = addPseudoHeader(Http2Headers.PseudoHeaderName.STATUS.value(), status, count);
+    count = addPseudoHeader(key1, value1, count);
+    psuedoHeadersCount = count;
+  }
+
+  private int addPseudoHeader(
+      AsciiString headerKey, @Nullable AsciiString headerValue, int currentCount) {
+    if (headerValue != null) {
```

Comment:
Also check the key since `key1`, 'key2`, and `key3` are nullable in the constructor?